### PR TITLE
Use plop for components scaffolding

### PR DESCRIPTION
**Pull request by fesor**
What this for:

 - To simplify components generation. You could now use `some component` as component name and it will generate proper names for controllers, files and so on.
 - Allow to add custom templates for code generation in future. We could prepare some base templates like `some.factory.js` and `some.factory.spec.js` and many other stuff. With plop this kind of generators pretty much easier to maintain.
 - ~~One step to get rid of gulp completely.~~

Waiting for any feedback.

### CODE REVIEW

**Comment by fesor** [comment]
Blocked by https://github.com/amwmedia/node-plop/issues/25

**Comment by fesor** [comment]
@gdi2290 @Hendrixer Could you take a look?

p.s. Maybe I should also ping someone else since you guys are always busy?

**Comment by d3viant0ne** [comment]
screen shot 2016-11-28 at 1 58 12 pm

I would personally question the wisdom in replacing an extremely stable tool `gulp` with an alpha release of a less known, limited adoption tool.

**Comment by d3viant0ne** [comment]
I'm all for adopting and supporting emerging technologies. I am also always up from being able to remove gulp from any workflow.

Given this is a starter project, people may have very well extended the base gulp configuration and it's removal would be detrimental. 

The feedback you really need is from the people depending on / working with this starter. That or you could run the two in parallel for the time being, deprecate the gulp workflow and then see how the potential change is received by the community prior to removing gulp.

**Comment by fesor** [comment]
node-plop is just inner dependency of plop. Plop itself has 1.7.2 version and it pretty much stable.

>  replacing an extremely stable tool gulp with an alpha release of a less known, limited adoption tool.

This PR aims to replace gulp-rename + gulp-template (which uses limited lodash templates) on specialized tool for code generation, which uses highly stable handlebars for templates and pretty much popular Inquirer.js for wizards. This allows to simplify boilerplate process (for example you could use component name "my card" instead of only "myCard". Plop will use proper case. So files would be named like "my-card.component.js" and module name will be "ng6Starter.myCard". Also it is pretty much easy to add your own templates of some other stuff.

My team working with plop for about 3 month and they find it much more flexible than just gulp-template + gulp-rename. 

> deprecate the gulp workflow

This PR doesn't removes gulp or gulp workflow, it just replaces `gulp component` command to more appropriate and flexible way of boilerplate generation. It has proper documentation and examples on how to make your own templates to reduce routine. 

Right now gulpfile is bloated with unneeded stuff which just confuses users. It only used to start webpack and browserify. And as for e2e tests gulp brings more problem than it solves. But removing gulp is another story and should not be discussed in context of this PR. I will create separate issue for that discussion.

Thanks.

**Comment by fesor** [comment]
@d3viant0ne I will be very appreciated if you could put some more input on this. With this PR I tried to solve this issues:

 - Make component templates to fit angular styleguide.
 - proper module name generation (I really not sure that this is best for now)
 - Simple way to add custom templates (currently adding new templates with gulp isn't something obvious, and plop has documentation and lots of examples)
 - ability to patch existing code (not included in PR) in order to register newly generated components.

File: config/plop.config.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+      {
+        type: 'input',
+        name: 'name',
+        message: 'What is name of new component?',
+        validate: required('name'),
+      },
+      {
+        type: 'confirm',
+        name: 'needPath',
+        message: 'Should I place new component in components directory, or it should be placed somewhere else?',
```
Comment (jrencz) [inline config/plop.config.js]: I'd recommend `Do you want to specify the path manually (otherwise component will be placed in \`components\`) `

**Comment by thematho** [comment]
I have the working code from this pull request ready to add also a service plop generator with Unit test and working example. Can we add this pull request into it's own branch so I can sent pull request directly to the branch instead of @fesor 's fork?

**Comment by thematho** [comment]
@fesor I added a pull request for the plop branch adding the rest of the elements for scaffolding:
[Added Service/Factory/Directive/Filter Generators](https://github.com/AngularClass/NG6-starter/pull/224)

**Comment by fesor** [comment]
Closing in favor of #224

**Pull request closed without merging**